### DISCUSSION TITLE
Advice of byte buddy throw java.lang.ClassNotFoundException when using my own class in the custom instrumentation

### DISCUSSION THREAD

**Discussion by bgong-mdsol** [open-telemetry/opentelemetry-java-instrumentation#3275] [Q&A]
Hi all, 
I like to write my own instrumentation to add some attributes based on http request. I write a helper class to build additional attributes, and get the exception **java.lang.ClassNotFoundException** when call it in Advice of byte buddy: 

    @Advice.OnMethodEnter(suppress = Throwable.class)
    public static void onEnter( 
        @Advice.Argument(0) ServletRequest request,
        @Advice.Argument(1) ServletResponse response) {
        if (!(request instanceof HttpServletRequest) || !(response instanceof HttpServletResponse)) {
           return;
        }
        try {
           AttributesHelper.testing((HttpServletRequest) request);
        } catch (Throwable t) {
           System.out.println("ERROR: " + t.getCause());
        }
    }
   
    public static class AttributesHelper { 
        public static void testing(HttpServletRequest httpRequest) {
            System.out.println("This is AttributesHelper");
        }
    }

When I run my app with the custom otel javaagent, got the error below, any suggestion?  
`ERROR: java.lang.ClassNotFoundException: my.javaagent.instrumentation.servlet.AttributesHelper`

**Comment by iNikem** [comment]
Maybe this will hetlp: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md#ishelperclass ?

**Reply by bgong-mdsol** [reply]
Thanks @iNikem, 
I refer the [doc](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md#ishelperclass) and [example](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/distro/instrumentation) to create my instrumentation, But I can't use my own class in **@Advice-annotated methods** because of **java.lang.ClassNotFoundException** .

I looked the existing instrumentation which is calling the method from different class in in **@Advice-annotated methods** , such as https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-scheduling-3.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/scheduling/TaskInstrumentation.java#L35

I am wonder if I missed any configuration or runtime dependence in my gradle project.

**Reply by iNikem** [reply]
@mateuszrzeszutek Can you help here?

**Reply by mateuszrzeszutek** [reply]
Hey @bgong-mdsol ,

Are you writing your instrumentation in a fork of this project or a distro/extension/repackage? The `isHelperClass(String)` method mentioned by @iNikem is called by our compile-time "muzzle" extension, which is not yet available for usage outside of this project.

As a workaround, please try manually adding all names of the classes that you use to the `getMuzzleHelperClassNames()` method, e.g.:

```java
public List<String> getMuzzleHelperClassNames() {
  return Collections.singletonList("my.javaagent.instrumentation.servlet.AttributesHelper");
}
```

This method is normally (well, normally in this project) generated automatically by the muzzle compile-time plugin, but if you don't use it you might have to write it yourself.

**Reply by bgong-mdsol** [reply]
👍  It works! Thanks a lot @mateuszrzeszutek and @iNikem
